# Python beginners course

The *pandas* package is the most important tool for Data Scientists and Analysts working in Python today. The powerful machine learning and glamorous visualization tools may get all the attention, but pandas is the backbone of most data projects. 

>\[*pandas*\] is derived from the term "**pan**el **da**ta", an econometrics term for data sets that include observations over multiple time periods for the same individuals. — [Wikipedia](https://en.wikipedia.org/wiki/Pandas_%28software%29)

If you're thinking about data science as a career, then it is imperative that one of the first things you do is learn pandas. In this course, we will go over the essential bits of information about pandas

## What's Pandas for?

Pandas has so many uses that it might make sense to list the things it can't do instead of what it can do. 

This tool is essentially your data’s home. Through pandas, you get acquainted with your data by cleaning, transforming, and analyzing it. 

For example, say you want to explore a dataset stored in a CSV / Excel file on your computer. Pandas will extract the data from that CSV into a DataFrame — a table, basically — then let you do things like:

- Calculate statistics and answer questions about the data, like


    - What's the average, median, max, or min of each column? 
    - Does column A correlate with column B?
    - What does the distribution of data in column C look like?


- Clean the data by doing things like removing missing values and filtering rows or columns by some criteria


- Visualize the data with help from Matplotlib. Plot bars, lines, histograms, bubbles, and more. 


- Store the cleaned, transformed data back into a CSV, other file or database



## How does pandas fit into the data science toolkit?

Not only is the pandas library a central component of the data science toolkit but it is used in conjunction with other libraries in that collection. 

Pandas is built on top of the **NumPy** package, meaning a lot of the structure of NumPy is used or replicated in Pandas. Data in pandas is often used to feed statistical analysis in **SciPy**, plotting functions from **Matplotlib**, and machine learning algorithms in **Scikit-learn**.

Jupyter Notebooks offer a good environment for using pandas to do data exploration and modeling, but pandas can also be used in text editors just as easily.

Jupyter Notebooks give us the ability to execute code in a particular cell as opposed to running the entire file. This saves a lot of time when working with large datasets and complex transformations. Notebooks also provide an easy way to visualize pandas’ DataFrames and plots. As a matter of fact, this article was created entirely in a Jupyter Notebook.

In [16]:
#import the pandas packages as 'pd'
import pandas as pd

## Core components of pandas: Series and DataFrames

The primary two components of pandas are the `Series` and `DataFrame`. 

A `Series` is essentially a column, and a `DataFrame` is a multi-dimensional table made up of a collection of Series. 

<img src="assets/series-and-dataframe.png" width=600px />

DataFrames and Series are quite similar in that many operations that you can do with one you can do with the other, such as filling in null values and calculating the mean.

You'll see how these components work when we start working with data below. 

### Creating DataFrames from scratch
Creating DataFrames right in Python is good to know and quite useful when testing new methods and functions you find in the pandas docs.

There are *many* ways to create a DataFrame from scratch, but a great option is to just use a simple `dict` (dictionary). A dictionary is a mapping of unique keys to values. 

Let's say we have a fruit stand that sells apples and oranges. We want to have a column for each fruit and a row for each customer purchase. To organize this as a dictionary for pandas we could do something like:

In [21]:
data = {
    'names': ['June', 'Robert', 'Lily', 'David'],
    'apples': [3, 2, 0, 1], 
    'oranges': [0, 3, 7, 2]
}

In the example above, *apples* and *oranges* are called the **keys** of the dictionary, while the lists on the righthandside are called **values** of the dictionary.

Now, we convert the dictionary into a DataFrame called *purchases*:

In [22]:
purchases = pd.DataFrame(data)

purchases

names  apples  oranges
0    June       3        0
1  Robert       2        3
2    Lily       0        7
3   David       1        2

**How did that work?**

Each *(key, value)* item in the dictionary corresponds to a *column* in the resulting DataFrame.

The **Index** of this DataFrame was given to us on creation as the numbers 0 to 3, but we could also create our own when we initialize the DataFrame. 

Let's have customer names as our index: 

In [28]:
purchases = pd.DataFrame(data)
purchases = purchases.set_index('names')

purchases

apples  oranges
names                  
June         3        0
Robert       2        3
Lily         0        7
David        1        2

So now we could **loc**ate a customer's order by using their name:

In [29]:
purchases.loc['June']

apples     3
oranges    0
Name: June, dtype: int64

There's more on locating and extracting data from the DataFrame later, but now you should be able to create a DataFrame with any random data to learn on.

Let's move on to some quick methods for creating DataFrames from various other sources.

## Exercise 1
Above we have seen how to manually create a DataFrame. Now lets try ourselves! Complete the steps below by filling in the ___.

### Step 1
Create a DataFrame that contains the following columns:
1. Column "Account holder" containing the names of bank account holders
2. Column "Account number" containing bank account numbers
3. Column "Balance" containing the current balance

### Step 2
Set the first column as the index column.

### Step 3
Locate the account number and balance of one of the account holders.



In [ ]:
#step 1
data = {
    ___
}

bankaccounts = pd.DataFrame(___)

#step 2
bankaccounts = bankaccounts.___

#step 3
located_account = bankaccounts.loc___

located_account

## How to read in data

Manually typing out data is nice, but we actually want to read in from spreadsheets. In the following examples we'll keep using our apples and oranges data, but this time it's coming from a CSV file. With CSV files (Excel-like files) all you need is a single line to load in the data:

In [31]:
purchases = pd.read_csv('assets/purchases.csv')

purchases

names  apples  oranges
0    June       3        0
1  Robert       2        3
2    Lily       0        7
3   David       1        2

Again we want the names to be used as an index. We can designate a column to be used as index using `index_col`:

In [6]:
df = pd.read_csv('assets/purchases.csv', index_col=0)

df

apples  oranges
June         3        0
Robert       2        3
Lily         0        7
David        1        2

Here we're setting the index to be column zero (NOTE: in Python, counting items usually starts at zero).

You'll find that most CSVs won't ever have an index column. In that case, the numbers assigned by default (0 to 3) are just fine to work with.

## Most important DataFrame operations
Now that we have a basic idea of how to load data into DataFrames, let's move on to importing some real-world data and detailing a few of the operations you'll be using a lot.

DataFrames possess hundreds of methods and other operations that are crucial to any analysis. As a beginner, you should know the operations that perform simple transformations of your data and those that provide fundamental statistical analysis.

In this training, we use the Boston Housing dataset which contains information collected by the U.S Census Service concerning housing in the area of Boston Mass. The dataset is relatively small in size with only 506 cases.

The data was originally published by Harrison, D. and Rubinfeld, D.L. Hedonic prices and the demand for clean air, J. Environ. Economics & Management, vol.5, 81-102, 1978.

It has four columns:
- CRIM - per capita crime rate by town
- ZN - proportion of residential land zoned for lots over 25,000 sq.ft.
- INDUS - proportion of non-retail business acres per town.
- CHAS - Charles River dummy variable (1 if tract bounds river; 0 otherwise)
- NOX - nitric oxides concentration (parts per 10 million)
- RM - average number of rooms per dwelling
- AGE - proportion of owner-occupied units built prior to 1940
- DIS - weighted distances to five Boston employment centres
- RAD - index of accessibility to radial highways
- TAX - full-value property-tax rate per 10,000 dollar
- PTRATIO - pupil-teacher ratio by town
- B - 1000(Bk - 0.63)^2 where Bk is the proportion of blacks by town
- LSTAT - percentage lower status of the population
- MEDV - Median value of owner-occupied homes (1 = 1000 dollar)


In [5]:
boston = pd.read_csv("data/boston_dataset.csv")
boston.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD  TAX  PTRATIO  \
0  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296     15.3   
1  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242     17.8   
2  0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242     17.8   
3  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222     18.7   
4  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222     18.7   

        B  LSTAT  MEDV  
0  396.90   4.98  24.0  
1  396.90   9.14  21.6  
2  392.83   4.03  34.7  
3  394.63   2.94  33.4  
4  396.90   5.33  36.2

We're loading this dataset from a CSV and designating the index column to be our index.

### Viewing your data

The first thing to do when opening a new dataset is print out a few rows to keep as a visual reference. We accomplish this with `.head()`:

In [11]:
boston.head()

LSTAT     RM  MEDV
0   4.98  6.575  24.0
1   9.14  6.421  21.6
2   4.03  7.185  34.7
3   2.94  6.998  33.4
4   5.33  7.147  36.2

`.head()` outputs the **first** five rows of your DataFrame by default, but we could also pass a number as well: `boston.head(10)` would output the top ten rows, for example. 

To see the **last** five rows use `.tail()`. `tail()` also accepts a number, and in this case we printing the bottom two rows.:

In [12]:
boston.tail(2)

LSTAT     RM  MEDV
504   6.48  6.794  22.0
505   7.88  6.030  11.9

Typically when we load in a dataset, we like to view the first five or so rows to see what's under the hood. Here we can see the names of each column, the index, and examples of values in each row.

### Getting info about your data

`.info()` should be one of the very first commands you run after loading your data:

In [13]:
boston.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 506 entries, 0 to 505
Data columns (total 3 columns):
LSTAT    506 non-null float64
RM       506 non-null float64
MEDV     506 non-null float64
dtypes: float64(3)
memory usage: 15.8 KB


`.info()` provides the essential details about your dataset, such as the number of rows and columns, the number of non-null values, what type of data is in each column, and how much memory your DataFrame is using. 

Another fast and useful attribute is `.shape`, which outputs just a tuple of (rows, columns):

In [14]:
boston.shape

(506, 3)

Note that `.shape` has no parentheses and is a simple tuple of format (rows, columns). So we have **506 rows** and **3 columns** in our DataFrame.

You'll be going to `.shape` a lot when cleaning and transforming data. For example, you might filter some rows based on some criteria and then want to know quickly how many rows were removed.

### Handling duplicates

This dataset does not have duplicate rows, but it is always important to verify you aren't aggregating duplicate rows. 

To demonstrate, let's simply just double up our DataFrame by appending it to itself:

In [15]:
temp_df = boston.append(movies_df)

temp_df.shape

(1012, 3)

Using `append()` will return a copy without affecting the original DataFrame. We are capturing this copy in `temp` so we aren't working with the real data.

Notice call `.shape` quickly proves our DataFrame rows have doubled.

Now we can try dropping duplicates:

In [16]:
temp_df = temp_df.drop_duplicates()

temp_df.shape

(506, 3)

Just like `append()`, the `drop_duplicates()` method will also return a copy of your DataFrame, but this time with duplicates removed. Calling `.shape` confirms we're back to the 1000 rows of our original dataset.

### Column renaming 

Many times datasets will have unpractical column names with symbols, upper and lowercase words, spaces, and typos. To make selecting data by column name easier we can spend a little time cleaning up their names.

Here's how to print the column names of our dataset:

In [19]:
boston.columns

Index(['LSTAT', 'RM', 'MEDV'], dtype='object')

Now, we can use the `.rename()` method to rename certain or all columns via a `dict`. For example, we want to rename `MEDV` to `MEDIAN` and `RM` to `ROOMS`:

In [21]:
boston = boston.rename(columns={
        'MEDV': 'MEDIAN', 
        'RM': 'ROOMS'
    })

boston.columns

Index(['LSTAT', 'ROOMS', 'MEDIAN'], dtype='object')

and then we change it back again:

In [22]:
boston = boston.rename(columns={
        'MEDIAN' :  'MEDV', 
        'ROOMS'  :  'RM'
    })

boston.columns

Index(['LSTAT', 'RM', 'MEDV'], dtype='object')

### Understanding your variables

To get an idea of the data you are working with, `describe()` can be used on an entire DataFrame to get a summary of the distribution of continuous variables:

In [32]:
boston.describe()

CRIM          ZN       INDUS        CHAS         NOX          RM  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean     3.613524   11.363636   11.136779    0.069170    0.554695    6.284634   
std      8.601545   23.322453    6.860353    0.253994    0.115878    0.702617   
min      0.006320    0.000000    0.460000    0.000000    0.385000    3.561000   
25%      0.082045    0.000000    5.190000    0.000000    0.449000    5.885500   
50%      0.256510    0.000000    9.690000    0.000000    0.538000    6.208500   
75%      3.677082   12.500000   18.100000    0.000000    0.624000    6.623500   
max     88.976200  100.000000   27.740000    1.000000    0.871000    8.780000   

              AGE         DIS         RAD         TAX     PTRATIO           B  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean    68.574901    3.795043    9.549407  408.237154   18.455534  356.674032   
std     28.148861    2.105710    8.707259  168.537116    2.164946   91.294864   
min      2.900000    1.129600    1.000000  187.000000   12.600000    0.320000   
25%     45.025000    2.100175    4.000000  279.000000   17.400000  375.377500   
50%     77.500000    3.207450    5.000000  330.000000   19.050000  391.440000   
75%     94.075000    5.188425   24.000000  666.000000   20.200000  396.225000   
max    100.000000   12.126500   24.000000  711.000000   22.000000  396.900000   

            LSTAT        MEDV  
count  506.000000  506.000000  
mean    12.653063   22.532806  
std      7.141062    9.197104  
min      1.730000    5.000000  
25%      6.950000   17.025000  
50%     11.360000   21.200000  
75%     16.955000   25.000000  
max     37.970000   50.000000

### Correlations

By using the correlation method `.corr()` we can generate the relationship between each continuous variable:

In [34]:
boston.corr()

LSTAT        RM      MEDV
LSTAT  1.000000 -0.613808 -0.737663
RM    -0.613808  1.000000  0.695360
MEDV  -0.737663  0.695360  1.000000

Correlation tables are a numerical representation of the bivariate relationships in the dataset. 

Positive numbers indicate a positive correlation — one goes up the other goes up — and negative numbers represent an inverse correlation — one goes up the other goes down. 1.0 indicates a perfect correlation. 

So looking in the first row, first column we see `LSTAT` has a perfect correlation with itself, which is obvious. On the other hand, the correlation between `RM` and `MEDV` is ~0.695. A little more interesting.

Examining bivariate relationships comes in handy when you have an outcome or dependent variable in mind and would like to see the features most correlated to the increase or decrease of the outcome. You can visually represent bivariate relationships with scatterplots (seen below in the plotting section). 

### DataFrame slicing, selecting, extracting

Up until now we've focused on some basic summaries of our data. Below are the other methods of slicing, selecting, and extracting you'll need to use constantly. Let's look at working with columns first.

#### Selecting a column
Using square brackets is the general way we select columns in a DataFrame:

In [7]:
medv_col = boston['MEDV']

type(medv_col)

pandas.core.series.Series

Notice that the selected column `medv_col` is not a DataFrame like `boston` is. Instead, `medv_col` is what we call a *Series*. It's important to note that, although many methods are the same, DataFrames and Series have different attributes, so you'll need be sure to know which type you are working with or else you will receive attribute errors. 

To extract a column as a *DataFrame* instead, you use square brackets to pass a list of column names. In our case that's just a single column:

In [8]:
medv_col = boston[['MEDV']]

type(medv_col)

pandas.core.frame.DataFrame

Since it's just a list, adding another column name is easy:

In [9]:
subset = boston[['RM', 'MEDV']]

subset.head()

RM  MEDV
0  6.575  24.0
1  6.421  21.6
2  7.185  34.7
3  6.998  33.4
4  7.147  36.2

Now we'll look at getting data by rows.

#### Selecting a row

For rows, we can use the `.iloc` method to **loc**ate by numerical **i**ndex. So to get the row at index 222 we do:

In [6]:
row = boston.loc[222]

row

CRIM         0.62356
ZN           0.00000
INDUS        6.20000
CHAS         1.00000
NOX          0.50700
RM           6.87900
AGE         77.70000
DIS          3.27210
RAD          8.00000
TAX        307.00000
PTRATIO     17.40000
B          390.39000
LSTAT        9.93000
MEDV        27.50000
Name: 222, dtype: float64

We can also use a trick called *slicing* to select multiple rows. Slicing is done using square brackets like `boston[1:4]`:

In [10]:
boston_subset = boston.iloc[20:30]

boston_subset

CRIM   ZN  INDUS  CHAS    NOX     RM    AGE     DIS  RAD  TAX  PTRATIO  \
20  1.25179  0.0   8.14     0  0.538  5.570   98.1  3.7979    4  307     21.0   
21  0.85204  0.0   8.14     0  0.538  5.965   89.2  4.0123    4  307     21.0   
22  1.23247  0.0   8.14     0  0.538  6.142   91.7  3.9769    4  307     21.0   
23  0.98843  0.0   8.14     0  0.538  5.813  100.0  4.0952    4  307     21.0   
24  0.75026  0.0   8.14     0  0.538  5.924   94.1  4.3996    4  307     21.0   
25  0.84054  0.0   8.14     0  0.538  5.599   85.7  4.4546    4  307     21.0   
26  0.67191  0.0   8.14     0  0.538  5.813   90.3  4.6820    4  307     21.0   
27  0.95577  0.0   8.14     0  0.538  6.047   88.8  4.4534    4  307     21.0   
28  0.77299  0.0   8.14     0  0.538  6.495   94.4  4.4547    4  307     21.0   
29  1.00245  0.0   8.14     0  0.538  6.674   87.3  4.2390    4  307     21.0   

         B  LSTAT  MEDV  
20  376.57  21.02  13.6  
21  392.53  13.83  19.6  
22  396.90  18.72  15.2  
23  394.54  19.88  14.5  
24  394.33  16.30  15.6  
25  303.42  16.51  13.9  
26  376.88  14.81  16.6  
27  306.38  17.28  14.8  
28  387.94  12.80  18.4  
29  380.23  11.98  21.0

#### Conditional selections
We’ve gone over how to select columns and rows, but what if we want to make a conditional selection? 

For example, what if we want to filter our DataFrame to show only item where the nitric oxides concentration exeeds 0.4 parts per million?

To do that, we take a column from the DataFrame and apply a Boolean condition to it. Here's an example of a Boolean condition:

In [12]:
condition = (boston['NOX'] > 0.4)

condition.head()

0    True
1    True
2    True
3    True
4    True
Name: NOX, dtype: bool

The Boolean condition above returns a Series of True and False values: True for concentrations above 0.4 and False for concentrations below. 

We want to filter out concentrations below 0.4, in other words, we don’t want rows that are False. To return the rows where that condition is True we have to pass this operation into the DataFrame:

In [14]:
boston[boston['NOX'] > 0.4].head(10)

CRIM    ZN  INDUS  CHAS    NOX     RM    AGE     DIS  RAD  TAX  PTRATIO  \
0  0.00632  18.0   2.31     0  0.538  6.575   65.2  4.0900    1  296     15.3   
1  0.02731   0.0   7.07     0  0.469  6.421   78.9  4.9671    2  242     17.8   
2  0.02729   0.0   7.07     0  0.469  7.185   61.1  4.9671    2  242     17.8   
3  0.03237   0.0   2.18     0  0.458  6.998   45.8  6.0622    3  222     18.7   
4  0.06905   0.0   2.18     0  0.458  7.147   54.2  6.0622    3  222     18.7   
5  0.02985   0.0   2.18     0  0.458  6.430   58.7  6.0622    3  222     18.7   
6  0.08829  12.5   7.87     0  0.524  6.012   66.6  5.5605    5  311     15.2   
7  0.14455  12.5   7.87     0  0.524  6.172   96.1  5.9505    5  311     15.2   
8  0.21124  12.5   7.87     0  0.524  5.631  100.0  6.0821    5  311     15.2   
9  0.17004  12.5   7.87     0  0.524  6.004   85.9  6.5921    5  311     15.2   

        B  LSTAT  MEDV  
0  396.90   4.98  24.0  
1  396.90   9.14  21.6  
2  392.83   4.03  34.7  
3  394.63   2.94  33.4  
4  396.90   5.33  36.2  
5  394.12   5.21  28.7  
6  395.60  12.43  22.9  
7  396.90  19.15  27.1  
8  386.63  29.93  16.5  
9  386.71  17.10  18.9

We can also make some richer conditionals by using logical operators `|` for "or" and `&` for "and".

Let's filter the the DataFrame to show only items where NOX is greater than 0.4 OR where TAX is lower than 300:

In [15]:
boston[(boston['NOX'] > 0.4) | (boston['TAX'] < 300)].head(10)

CRIM    ZN  INDUS  CHAS    NOX     RM    AGE     DIS  RAD  TAX  PTRATIO  \
0  0.00632  18.0   2.31     0  0.538  6.575   65.2  4.0900    1  296     15.3   
1  0.02731   0.0   7.07     0  0.469  6.421   78.9  4.9671    2  242     17.8   
2  0.02729   0.0   7.07     0  0.469  7.185   61.1  4.9671    2  242     17.8   
3  0.03237   0.0   2.18     0  0.458  6.998   45.8  6.0622    3  222     18.7   
4  0.06905   0.0   2.18     0  0.458  7.147   54.2  6.0622    3  222     18.7   
5  0.02985   0.0   2.18     0  0.458  6.430   58.7  6.0622    3  222     18.7   
6  0.08829  12.5   7.87     0  0.524  6.012   66.6  5.5605    5  311     15.2   
7  0.14455  12.5   7.87     0  0.524  6.172   96.1  5.9505    5  311     15.2   
8  0.21124  12.5   7.87     0  0.524  5.631  100.0  6.0821    5  311     15.2   
9  0.17004  12.5   7.87     0  0.524  6.004   85.9  6.5921    5  311     15.2   

        B  LSTAT  MEDV  
0  396.90   4.98  24.0  
1  396.90   9.14  21.6  
2  392.83   4.03  34.7  
3  394.63   2.94  33.4  
4  396.90   5.33  36.2  
5  394.12   5.21  28.7  
6  395.60  12.43  22.9  
7  396.90  19.15  27.1  
8  386.63  29.93  16.5  
9  386.71  17.10  18.9

As you can see, we can easily filter our data to get only the values we are interested in using Boolean conditions.

## Wrapping up

Exploring, cleaning, transforming, and visualization data with pandas in Python is an essential skill in data science. Just cleaning wrangling data is 80% of your job as a Data Scientist. After a few projects and some practice, you should be very comfortable with most of the basics.

To keep improving, view the [extensive tutorials](https://pandas.pydata.org/pandas-docs/stable/tutorials.html) offered by the official pandas docs, follow along with a few [Kaggle kernels](https://www.kaggle.com/kernels), and keep working on your own projects!

Moreover, for those of you looking to do a [data science bootcamp](https://www.learndatasci.com/articles/thinkful-data-science-online-bootcamp-review/) or some other accelerated data science education program, it's highly recommended you start learning pandas on your own before you start the program. 